src/transformers/models/chameleon/logits_processors.py

```diff
@@ -0,0 +1,306 @@
+"""
+[Chameleon's generator](https://github.com/facebookresearch/chameleon/blob/main/chameleon/inference/chameleon.py#L374) uses a finite state machine (FSM) with the modality of the tokens as states to dynamically switch between text and image generation. The generator uses this FSM to determine which tokens are allowed to be generated next based on the previously generated tokens. Additionally, due to the limitations of the VQ-VAE model used to tokenize & detokenize images, the generator also ensures that:
+
+    1. The image tokens are generated in a contiguous block of `image_seq_length` tokens, which is currently fixed at 1024.
+    2. The block of image tokens is preceded by the `image_start_token` (`<racm3:break>`) and followed by the `image_end_token` (`<eoss>`) marker tokens. And that
+    3. The block of image tokens doesn't overshoot the maximum sequence length (`max_length`). It does this by simply preventing the `image_start_token` from being generated if there isn't enough space for the rest of the tokens.
+
+Chameleon is so dependent on this FSM that removing it leads to sever model degradation. Removing or ignoring the invalid/nonsensical tokens as they are generated is not enough--we _have_ to mask their logits out entirely.
+
+Here, we implement four generation modes, each with its own FSM guide (or none thereof):
+
+    1. `text-only` (default): We mask out the logits of image tokens. The `image_start_token` and `image_end_token` are allowed but they have to close each other out immediately. This is done by the `ChameleonTextOnlyLogitsProcessor` class.
+    2. `image-only`: We mask out the logits of non-image tokens.
+    3. `interleaved-text-image`: We allow interleaved generation of text and images. This is actually the default mode of the official Chameleon model. And
```

Comment:
Thanks for a detailed explanation. This actually belongs in the `transformers/generation/logits_processors.py`. 

I am not sure about the general design of FSM however, as it is a lot of code to maintain and is tailored for Chameleon case only. I remember we had some discussions of making outlines part of the library. Ping @gante on this

Reply:
Custom generation code is extremely costly to maintain given our current bandwidth, and should only be done in very high impact models (like `whisper`). As such, my suggestion here is to NOT add this generation code to `transformers`.

We can, however, add it to non-actively maintained parts of the library. For instance, it could go into the `examples/research_projects` folder, where we can pin a `transformers` version and don't worry about maintenance. We can also link to that research project on the model card, for further visibility.

If this FSM pattern pops up in more VLMs, we can then add it to the main `transformers` body.

WDYT? 🤗

Reply:
We discussed this internally on slack. So, given the maintenance cost of FSM and probably not so wide usage beyond Chameleon, we want to find another way to support interleaved-generation.

@leloykun am i right that w/o this the FSM the generation completely fails or it's only about the quality of generation? If the second case, what's the difference in performance between the two versions?

Reply:
@zucchini-nlp, unfortunately, it seems that Chameleon can't do interleaved generation by itself without the FSM. It just produces gibberish & can't generate images properly. So yah, if we remove the FSM, we'll have to turn off interleaved generation.

But for the text-only & image-only modes, we'd just need to mask the logits for the other modality. However, I think we should do this outside of the _forward func and preferably w/ a LogitsProcessor for consistency

Reply:
Just chiming in here, you could implement this specific part in Outlines easily for now. The FSM logic is heavy on maintenance; the only viable alternative would be importing this from Outlines but that’s a longer discussion 🙂

Reply:
Hmm, in that case it would be better to support text-only & image-only models in the library. And add in the docs that for interleaved generation one can install and use `outlines` processors. 

> But for the text-only & image-only modes, we'd just need to mask the logits for the other modality. However, I think we should do this outside of the _forward func and preferably w/ a LogitsProcessor for consistency

Yes, totally agreed. Now that you've mentioned it, I remembered we have this processor that sets some ids to "-inf". We can use that for Chameleon

https://github.com/huggingface/transformers/blob/24cfcc2114b5daf6ca1cce6e838ee9aa17096e62/src/transformers/generation/logits_process.py#L1801